addons/mail/models/mail_activity.py

```diff
@@ -189,71 +190,71 @@ def _onchange_recommended_activity_type_id(self):
-            # create / read: just check rights on related document
-            activity_to_documents.setdefault(activity.res_model, list()).append(activity.res_id)
-        for doc_model, doc_ids in activity_to_documents.items():
-            if hasattr(self.env[doc_model], '_mail_post_access'):
-                doc_operation = self.env[doc_model]._mail_post_access
+        result = super()._check_access(operation)
+        if not self:
+            return result
+
+        # determine activities to check
+        if operation == 'read':
+            activities = self - self.browse(result)
+            activities -= activities.sudo().filtered_domain([('user_id', '=', self.env.uid)])
+        elif operation in ('write', 'unlink'):
+            activities = self.browse(result)
+            result = {}
+        else:
+            activities = self - self.browse(result)
+
+        if not activities:
+            return result
+
+        # now check access on related document of 'activities', and collect the
+        # ids of forbidden activities
+        model_docid_actids = defaultdict(lambda: defaultdict(list))
+        for activity in activities.sudo():
+            model_docid_actids[activity.res_model][activity.res_id].append(activity.id)
+
+        forbidden_ids = []
+        for doc_model, docid_actids in model_docid_actids.items():
+            documents = self.env[doc_model].browse(docid_actids)
+            if hasattr(documents, '_mail_post_access'):
+                doc_operation = documents._mail_post_access
             elif operation == 'read':
-                doc_operation = 'read'
+                doc_operation == 'read'
             else:
-                doc_operation = 'write'
-            right = self.env[doc_model].check_access_rights(doc_operation, raise_exception=False)
-            if right:
-                valid_doc_ids = getattr(self.env[doc_model].browse(doc_ids), filter_access_rules_method)(doc_operation)
-                valid += remaining.filtered(lambda activity: activity.res_model == doc_model and activity.res_id in valid_doc_ids.ids)
+                doc_operation == 'write'
+            for doc_id in documents._check_access(doc_operation):
+                forbidden_ids.extend(docid_actids[doc_id])
 
-        return valid
+        result.update(dict.fromkeys(
+            forbidden_ids,
+            lambda: self.browse(forbidden_ids)._make_access_error(operation)
+        ))
+        return result
+
+    def _make_access_error(self, operation: str) -> AccessError:
+        return AccessError(_(
+            "The requested operation cannot be completed due to security restrictions. "
+            "Please contact your system administrator.\n\n"
+            "(Document type: %(type)s, Operation: %(operation)s)\n\n"
+            "Records: %(records)s, User: %(user)s",
+            type=self._description,
+            operation=operation,
+            records=format_list(self.env, list(map(str, self.ids[:6]))),
```

Comment:
I would not use `format_list` in the exceptions. Just `", ".join` because format list add a translated "and" which is not needed in this kind of technical messages and adds a step when you copy ids.

Reply:
It's indeed technical information.  `format_list` would only make sense when showing the `display_name` of `self`.